### PR TITLE
set CMAKE_SYSTEM_NAME when cross compiling

### LINKED ISSUE #4550: cmake cross compilation does not set CMAKE_SYSTEM_NAME
[Documentation ](https://docs.conan.io/en/latest/reference/build_helpers/cmake.html?highlight=cmake_system_name#definitions)states that `CMAKE_SYSTEM_NAME` is `Set to self.settings.os value if cross-building is detected`, which is a good thing, because it will make [cmake set the `CMAKE_CROSSCOMPILING` variable automatically](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html)
However, when trying to cross compile [mysql-connector-c](https://github.com/bincrafters/conan-mysql-connector-c) (linux x86_64 to linux armv7hf) [the build fails](https://github.com/bincrafters/community/issues/644), because `CMAKE_SYSTEM_NAME` is not set by conan. It looks like conan sets the CMAKE_SYSTEM_NAME variable only if the the host os is not the build os, which is wrong: we can cross compile to the same os with a different architecture.

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [X] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by ericLemanissier**
fixes #4550 

- [X] Refer to the issue that supports this Pull Request.
- [X] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [X] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [X] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

File: conans/client/build/cmake_flags.py [hunk 1]
```diff
@@ -152,12 +152,11 @@ def _cmake_cross_build_defines(self):
...
-                    if os_:  # the_os is the host (regular setting)
-                        ret["CMAKE_SYSTEM_NAME"] = ("Darwin" if os_ in ["iOS", "tvOS", "watchOS"]
-                                                    else os_)
-                    else:
-                        ret["CMAKE_SYSTEM_NAME"] = "Generic"
+                if os_:  # the_os is the host (regular setting)
+                    ret["CMAKE_SYSTEM_NAME"] = ("Darwin" if os_ in ["iOS", "tvOS", "watchOS"]
+                                                else os_)
+                else:
+                    ret["CMAKE_SYSTEM_NAME"] = "Generic"
```
Comment (lasote) [inline conans/client/build/cmake_flags.py]: `cross_building` also returns True when x86_64 => x86 and I'm not sure we should adjust `CMAKE_SYSTEM_NAME` in that case.

Comment (lasote) [reply conans/client/build/cmake_flags.py]: Reading this: https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html probably you are right. But I'm almost sure this produce some other issue, I'm trying to recall and asking people.

Comment (ericLemanissier) [reply conans/client/build/cmake_flags.py]: browsing the history I'd say that this `if` should have been removed with https://github.com/conan-io/conan/commit/20eb12fe73c6ac2b9b2ab06a3fab684e83ae58a1#diff-029e992e62f056e56357fa0ba4cf4111R160, but it is true that this change can have some unexpected consequences on x86_64->x86 builds

Comment (lasote) [reply conans/client/build/cmake_flags.py]: Yeah, my commit message `Not adjust system name when the os is the same but the arch is not the same` is disturbing. I did it for some reason... but it is a shame I didn't referece any issue.

Comment (ericLemanissier) [reply conans/client/build/cmake_flags.py]: One of the thing which breaks with when cross compiling from x86_64->x86 on same os is [try_run](https://cmake.org/cmake/help/latest/command/try_run.html#behavior-when-cross-compiling). The current PR should handle it correctly

**Comment by ericLemanissier** [comment]
Explicitly checking for `arch_build != "x86_64" or arch != "x86"` feels too low level.
I already had to include an equivalent piece of code in a recipe :
```
    if (not tools.cross_building(self.settings)) or\
            (self.settings.os_build == self.settings.os and\
             self.settings.arch_build == "x86_64" and self.settings.arch == "x86"):
```
And I think most of the test recipes should start the `test` method with this check.
It feels dirty, especially with conan warning:
```
************************************************************
This package defines both 'os' and 'os_build' 
Please use 'os' for libraries and 'os_build'
only for build-requires used for cross-building
************************************************************
```
Maybe this information should be provided by conan with a method `tools.cross_building_and_not_able_to_run_natively()` ?

**Comment by ericLemanissier** [comment]
I updated the PR to use `skip_x64_x86` argument of `tools.cross_building()`, so my last message is not relevant anymore.

**Comment by jgsogo** [comment]
In https://github.com/conan-io/conan/pull/7827, this behavior was implemented, `CMAKE_SYSTEM_NAME` is assigned whenever there is a cross-compiling scenario (we keep the `skip_x64_x86=True` for Windows/Linux because otherwise it would break existing behavior).

I'm closing this PR as it is no longer needed, but thank you anyway 👍

**Pull request closed without merging**